src/java/org/apache/cassandra/service/disk/usage/DiskUsageMonitor.java

```diff
@@ -0,0 +1,171 @@
+import org.apache.cassandra.io.util.FileUtils;
+
+/**
+ * Schedule periodic task to monitor local disk usage and notify {@link DiskUsageBroadcaster} if local state changed.
+ */
+public class DiskUsageMonitor
+{
+    private static final Logger logger = LoggerFactory.getLogger(DiskUsageMonitor.class);
+
+    public static DiskUsageMonitor instance = new DiskUsageMonitor();
+
+    private final GuardrailsConfig config = Guardrails.CONFIG_PROVIDER.getOrCreate(null);
+    private final Supplier<Set<Directories.DataDirectory>> dataDirectoriesSupplier;
+
+    private volatile DiskUsageState localState = DiskUsageState.NOT_AVAILABLE;
+
+    @VisibleForTesting
+    public DiskUsageMonitor()
+    {
+        this.dataDirectoriesSupplier = Directories.dataDirectories::getAllDirectories;
+    }
+
+    @VisibleForTesting
+    public DiskUsageMonitor(Supplier<Set<Directories.DataDirectory>> dataDirectoriesSupplier)
+    {
+        this.dataDirectoriesSupplier = dataDirectoriesSupplier;
+    }
+
+    /**
+     * Start monitoring local disk usage and call notifier when local disk usage state changed.
+     */
+    public void start(Consumer<DiskUsageState> notifier)
+    {
+        // start the scheduler regardless guardrail is enabled, so we can enable it later without a restart
+        ScheduledExecutors.scheduledTasks.scheduleAtFixedRate(() -> {
+
+            if (!Guardrails.localDataDiskUsage.enabled(null))
+                return;
+
+            updateLocalState(getDiskUsage(), notifier);
+        }, 0, CassandraRelevantProperties.DISK_USAGE_MONITOR_INTERVAL_MS.getLong(), TimeUnit.MILLISECONDS);
+    }
+
+    @VisibleForTesting
+    public void updateLocalState(double usageRatio, Consumer<DiskUsageState> notifier)
+    {
+        double percentage = usageRatio * 100;
+        long percentageCeiling = (long) Math.ceil(percentage);
+
+        DiskUsageState state = getState(percentageCeiling);
+
+        Guardrails.localDataDiskUsage.guard(percentageCeiling, state.toString(), false, null);
+
+        // if state remains unchanged, no need to notify peers
+        if (state == localState)
+            return;
+
+        localState = state;
+        notifier.accept(state);
+    }
+
+    /**
+     * @return local node disk usage state
+     */
+    @VisibleForTesting
+    public DiskUsageState state()
+    {
+        return localState;
+    }
+
+    /**
+     * @return disk usage (including all memtable sizes) ratio
+     */
+    @VisibleForTesting
+    public double getDiskUsage()
+    {
+        // using BigDecimal to handle large file system
+        BigDecimal used = BigDecimal.ZERO;
+        BigDecimal total = BigDecimal.ZERO;
+
+        for (Directories.DataDirectory dir : dataDirectoriesSupplier.get())
+        {
+            used = used.add(BigDecimal.valueOf(dir.getSpaceUsed()));
+            total = total.add(BigDecimal.valueOf(dir.getTotalSpace()));
+        }
+
+        if (config.getDataDiskUsageMaxDiskSize().toGibibytes() > 0)
```

Comment:
I can not use anything less than 1GiB for `data_disk_usage_max_disk_size`. Is this what we want? I was trying to set it like to 200MiB and it will be rounded down to zero, so this will not be executed:

        if (config.getDataDiskUsageMaxDiskSize().toGibibytes() > 0)
        {
            long diskUsageMaxSizeInBytes = config.getDataDiskUsageMaxDiskSize().toBytes();
            total = total.min(BigDecimal.valueOf(diskUsageMaxSizeInBytes));
        }

Which means my total will be still total of the disk. For testing purposes, when I am trying to generate something quick in docker or whatever, megabytes are completely enough, no need to do gigabytes. I can we could lower it down to mebibytes maybe, on the other hand, it would be nice if we do not allow it to be super low, like 1MiB, at least 10MiB, because an empty Cassandra node, after bootstrapping, will occupy some space already.

An ideal solution would be to create a startup check for this - it would check if the value you have set makes sense - if it is acutally bigger than the stuff on the disk.

Reply:
Good catch, we should use `onfig.getDataDiskUsageMaxDiskSize().toBytes()` to avoid rounding problems. I'm also adding JMX methods to allow specifying different units than gibibytes.